Border Panel: Move global styles forcing of split borders to relevant panel's onChange

aaronrobertshaw opened the pull request:
Fixes: https://github.com/WordPress/gutenberg/issues/49594

## What?

Prior to https://github.com/WordPress/gutenberg/pull/48636, only the Global Styles would force a split definition of border styles. This PR moves that logic to the Global Styles' border panel's `onChange` thereby restoring the original behaviour of the individual block border supports.

## Why?

Forcing split borders results in bloated CSS as every longhand style is added individually to a block's inline styles. 

## How?

- Moves the forcing of split border definitions from  `onBorderChange` in the generic `BorderPanel` to the `onChange` of the Global Styles' `BorderPanel`.

## Testing Instructions

1. Before checking out this PR branch, edit a post and add a simple block supporting borders e.g. Group
2. Configure a simple "flat" border on the block, save, and view on the frontend
3. Note that the block will have separate longhand CSS rules inline for each side
4. Checkout this PR, repeat the process and note that there will only be an inline style for each property i.e. color, style, width etc, not for each property and style.


## Screenshots or screencast <!-- if applicable -->

| Before | After |
|---|---|
| <img width="1217" alt="Screenshot 2023-04-06 at 4 42 36 pm" src="https://user-images.githubusercontent.com/60436221/230295455-41d641ee-8f84-4a80-bbfa-ecd4ec91a227.png"> | <img width="1216" alt="Screenshot 2023-04-06 at 4 45 35 pm" src="https://user-images.githubusercontent.com/60436221/230295427-cfa72785-fbd7-4ab9-ad79-d7eae836f9a8.png"> |

aaronrobertshaw committed:
Fix typo throwing error

aaronrobertshaw committed:
Fix on border change handling in border panel

youknowriad commented on packages/block-editor/src/components/global-styles/border-panel.js:
```diff
@@ -187,38 +188,45 @@ export default function BorderPanel( {
-				color: encodeColorValue( updatedBorder[ side ]?.color ),
-			};
-		} );
+		if ( hasSplitBorders( updatedBorder ) ) {
```
So it's still possible to have split borders or not for the "block" case. In other words, the blocks are forced to understand the split borders, so why do we have an issue for the table block do you know?

aaronrobertshaw replied:
IIRC there was a decision made that setting differing borders per side on the Table block only made sense on the table as a whole not individual cells.

youknowriad replied:
Ok I confirmed my hunch. If I use split borders in the table block, the updates are not reflected on the UI, unless all the values are equal.

youknowriad replied:
Actually, they are reflected in the UI, it's just the 0 border that is wrong. So in other words, I think a better fix is to figure out why the 0 border is not reflected in the UI in the table block?

youknowriad replied:
If I remove the "has-individual-borders" class in the table block, things seem to start working properly, what am I missing there?

aaronrobertshaw replied:
> Actually, they are reflected in the UI, it's just the 0 border that is wrong. So in other words, I think a better fix is to figure out why the 0 border is not reflected in the UI in the table block?

The table is one issue. I recall there being some edge cases with the split border definitions and global styles when some things were defined as flat borders or not. I'm out of time today but would like to look into that history tomorrow to make sure that we don't introduce another regression.

youknowriad replied:
yeah, but If I'm not wrong, the component always uses split borders that way global styles continue to work. And blocks are supposed to understand split borders already. (you can set them in the UI), so what I'm saying is that there's a bug in the Table block when the user uses "split borders" regardless of my refactoring PR.

youknowriad replied:
Even in the current PR, you can reproduce the bug in the Table block, if you set 0 borders in some directions (left, right) but not all

aaronrobertshaw replied:
> the component always uses split borders

I'm not sure I'm following. I'll need to revisit this tomorrow to catch up. This conditional has an else clause handling the flat border's color slug to align with your changes in the panel refactor. When you edit a flat border on an individual block this panel gets a flat non-split border.

FWIW, previously, individual blocks could have a simple flat border that reduces the inline styles in the block markup. Now, after the panel refactor, split borders have been forced so that's always full longhand for each property. That's a step back right?

youknowriad replied:
I'm saying that on the block, you click the "split" icon and actually use split borders, and putting different values in each side forces the component to use the "long format". So the block needs to understand that format too regardless.

Here's the bug


https://user-images.githubusercontent.com/272444/230041506-5fda5971-97bd-4938-bdf9-2b3e4b1597ee.mov

As you can see, I can set split borders with random values like 10 and it works properly, but I can't set split borders with the 0 value. 0 only works if I use flat border.

aaronrobertshaw replied:
Thanks for explaining further 👍️
> i’m saying that on the block, you click the "split" icon and actually use split borders, and putting different values in each side forces the component to use the "long format".


Understood. I agree this PR isn’t a complete fix for the Table block. It appears as though if you define a subset of side definitions with zero width, they are being filtered out or not applied to the table. I’ll look into that one tomorrow as well.

As noted above, the table borders are just one example issue after the recent forcing of split borders.

There’s still the issue of the bloated CSS forcing separate longhand values creates. I’m not aware of the style engine having the ability to merge longhand styles to shorthand yet. I imagine that optimisation has been considered and is on the cards.

> So the block needs to understand that format too regardless.


I agree that the Table block needs some level of understanding of the split definitions. I don’t think that’s true for all blocks though. Most individual blocks are oblivious to whether they have long or shorthand definitions. The border object is a property on their style object that is processed by the style engine and they’re given whatever classes and styles needed. 

So to me, it’s here when processing the border change that we should be keeping a flat border configuration when possible. Which then allows the output and saving of minimal inline styles. There’s still the original edge case I need to dig up to explain what forced the global styles to rely on splitting the definitions to the longhand value in the first place. 

If we can extend the style engine to be able to handle converting longhand rules to shorthand where possible, I'd be all for forcing split border definitions across the board.

youknowriad replied:
Let's be clear I don't mind keeping the "flat" format when possible but I don't like the flag we pass to make an exception for global styles. I think the "style" object from global styles and blocks should be as similar as possible. (We're obviously not there in a lot of things). 

At the same time, I don't think using long format consistently (which my PR did) created any bug in the table block, the bug already existed, it was just manifesting in less cases and the fix to that bug is not really reverting to short syntax. The fix is closer to this https://github.com/WordPress/gutenberg/pull/49603?notification_referrer_id=NT_kwDOAAQoPLE2MDcxNDI2NzAwOjI3MjQ0NA#discussion_r1158232779

Now if you feel strongly that we should land the current PR as a stop gap to be back to the previous state. I won't block it because it's minor but I still feel that it's not ideal.

aaronrobertshaw replied:
I've moved the logic forcing split borders to the Global Styles' border panel's `onChange` avoiding the introduction of any new prop (https://github.com/WordPress/gutenberg/pull/49603/commits/854ab4be06c3cc21ebf8a13c6f8fd348c2314914).

The PR description has been updated to focus only on the return to a reduced number of CSS styles being inlined. I'm working on a fix for the table block separately, so I think we can move forward and land this one.

aaronrobertshaw committed:
Avoid using prop to modify border change behaviour

aaronrobertshaw committed:
Prevent encoding colors for sides that aren't set

Also prevents attempting to encode flat color for border object that is undefined.

youknowriad reviewed (approved):
LGTM 👍 Thanks for making these updates.
